crates/subspace-solving/src/lib.rs

```diff
@@ -42,12 +42,15 @@ mod construct_uint {
     }
 
     impl PieceDistance {
-        /// Calculates the xor distance metric between piece index hash and farmer address.
-        pub fn xor_distance(
-            PieceIndexHash(piece): &PieceIndexHash,
-            address: impl AsRef<[u8]>,
-        ) -> Self {
-            Self::from_big_endian(piece) ^ Self::from_big_endian(address.as_ref())
+        /// Calculates the distance metric between piece index hash and farmer address.
+        pub fn distance(PieceIndexHash(piece): &PieceIndexHash, address: impl AsRef<[u8]>) -> Self {
+            let piece = Self::from_big_endian(piece);
+            let address = Self::from_big_endian(address.as_ref());
+            if piece < address {
+                address - piece
+            } else {
+                piece - address
+            }
```

Comment:
This is not a correct algorithm, please port this algorithm instead: https://github.com/subspace/subspace/blob/f59154ae1cd7d2df9b233c3b77f108f1d82bc3d1/crates/sp-consensus-subspace/src/digests.rs#L40-L46

I'd like to have it in the `subspace-core-primitives` BTW.

For instance, it will produce incorrect result if `piece` is `2**256-1` and `address` is `0`. the distance should be `1` because they are next to each other across `0`, but here it'll be `2**256-1`.

Reply:
Can I generalize this logic (in terms of `num_traits`) and move it to some crate?

If yes, where it should be placed?

Reply:
I was thinking about something like this in `subspace-core-primitives`:
```rust
fn bidirectional_distance<T: WrappingSub + Min + Max>(a: T, b: T) -> T;
```

Not sure we need `num_traits` though, maybe we do :man_shrugging:

Reply:
Yeah, `WrappingSub` is in `num_traits`. I believe there is no trait for `Min` and `Max` if they are for minimum and maximum numbers. Overall we don't seem to even need them. Just this should do:
```rust
        let diff = self.wrapping_sub(other);
        let diff2 = other.wrapping_sub(self);
        // Find smaller diff between 2 directions.
        std::cmp::min(diff, diff2)
```

Reply:
Right, by `Min` I actually meant `Cmp` and for `Max` I meant the constant in code linked above.